src/python/pants/option/options_fingerprinter.py

```diff
@@ -22,21 +24,14 @@ class Encoder(json.JSONEncoder):
   def _encode_dict_wrapping_keys(self, d):
     return {self.encode(k):v for k, v in d.items()}
 
-  def encode(self, o):
-    if isinstance(o, (tuple, set)):
-      o = list(o)
-    elif isinstance(o, dict):
-      o = self._encode_dict_wrapping_keys(o)
-    elif len(o) == 1:
-      only_element = next(iter(o))
-      if isinstance(only_element, dict):
-        o = [self._encode_dict_wrapping_keys(only_element)]
-
-    return super(Encoder, self).encode(o)
-
   def default(self, o):
     if o is UnsetBool:
       return '_UNSET_BOOL_ENCODING'
+    elif isinstance(o, collections.Iterable):
+      if isinstance(o, collections.Mapping):
+        return self._encode_dict_wrapping_keys(o)
```

Comment:
Why? The point of `default` is it is called recursively for values. IE: if the encoder hits a dict with a key or a value it does not understand it tries default. As mentioned above - without this change, tests pass.

Reply:
That makes sense, and is very different from the understanding I had a few seconds ago. This is still failing for the test case I just added: `self.assertEqual(stable_json_sha1([{('a', 2): (3, 4)}]), '???')` (raising `TypeError: key (u'a', 2) is not a string`). I will read up on the json encoder docs and make this pass.

Reply:
"This" meaning "the right thing to do, not what I just pushed".

Reply:
OK - gotcha. Yeah, JSON objects can only have string keys, so special handling is needed.